sklearn/decomposition/dict_learning.py

```diff
@@ -309,19 +309,45 @@ def sparse_encode(X, dictionary, gram=None, cov=None, algorithm='lasso_lars',
     return code
 
 
-def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
-                 random_state=None):
-    """Update the dense dictionary factor in place.
+def _proj_l2(v, copy=False):
+    """Projects v unto l2 unit ball in-place.
+    """
+    if copy:
+        v = v.copy()
+    vv = np.dot(v, v)
+    if vv > 1.:
+        v /= sqrt(vv)
+    return v
+
+
+def _update_dict(dictionary, B, A, verbose=False, return_r2=False,
+                 random_state=None, online=False):
+    """Update the dense dictionary factor in place. It does a pass of BCD
+    to update by minimizing
+
+    .5 * ||Xt - DC||_F^2 = .5 * tr(DtDA) - 2 * tr(DtB) + tr(XXt),
+
+    as a function of the dictionary (D), where
+
+    B = XtCt, A = CCt, R = B - DA.
+
+    Note that the update of the jth atom is given by (see eqn 10 of
+    ref paper below):
+
+    R = R + D[:, k]A[k, :]  # rank-1 update
+    D[:, k] = R[:, k] / A[k, k]  # = D[:, k] + R[k] / A[k, k]
+    D[:, k] = proj(D[:, k])  # eqn 10
+    R = R - D[:, k]A[k, :]  # rank-1 update
```

Comment:
raghav makes a good point, we need to document A and B. No, A are not atoms, D[:, k] are the atoms.